### PR TITLE
Add GPU version of contrib.boolean_mask

### PR DESCRIPTION
**Pull request by HyperZealot**
## Description ##
The same as title

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- With additional parallelization of CPU version

### CODE REVIEW

**Comment by junrushao1994** [comment]
Please fix the CI issues :-)

File: src/operator/contrib/boolean_mask.cu [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  int32_t valid_num = 0;
+  int32_t* prefix_sum = nullptr;
+  void* d_temp_storage = nullptr;
+  size_t temp_storage_bytes = 0;
+  cub::DeviceScan::InclusiveSum(d_temp_storage,
+                                temp_storage_bytes,
+                                prefix_sum,
+                                prefix_sum,
+                                idx_size,
+                                Stream<gpu>::GetStream(s));
```
Comment (zheng-da) [inline src/operator/contrib/boolean_mask.cu]: could you please add comments to explain the purpose of the code?

**Comment by zheng-da** [comment]
great work! thanks

**Comment by vandanavk** [comment]
@mxnet-label-bot add [pr-awaiting-review, Operator]

File: src/operator/contrib/boolean_mask.cu (hunk 1 above)
Comment (HyperZealot) [reply src/operator/contrib/boolean_mask.cu]: Done, please check.

File: src/operator/contrib/boolean_mask.cc [hunk 2]
```diff
@@ -75,9 +74,118 @@ bool BooleanMaskBackStorageType(const nnvm::NodeAttrs& attrs,
...
+};
+
+template<>
+inline void BooleanMaskForward<cpu>(const nnvm::NodeAttrs& attrs,
+                                    const OpContext &ctx,
+                                    const std::vector<NDArray> &inputs,
+                                    const std::vector<OpReqType> &req,
+                                    const std::vector<NDArray> &outputs) {
+  // TODO(@junrushao1994): This implementation is a proof-of-concept,
+  // hence very slow actually. Performance should be improved in the future.
```
Comment (zheng-da) [inline src/operator/contrib/boolean_mask.cc]: please remove the comment here.

Comment (HyperZealot) [reply src/operator/contrib/boolean_mask.cc]: Done.

**Comment by HyperZealot** [comment]
@zheng-da  Should be good for merge once build passes

**Comment by junrushao1994** [comment]
Looks like there are some issues with the CI. Could you try again?

**Pull request merged**